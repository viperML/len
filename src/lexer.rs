use chumsky::extra::ParserExtra;
use chumsky::input::StrInput;

use chumsky::text::Char;

use crate::Int;
use chumsky::prelude::*;
use chumsky::Parser;

#[derive(Debug, Clone, PartialEq, PartialOrd)]
pub enum TokenKind<'src> {
    Number(Int),
    Bind,
    String(&'src str),
    Ident(&'src str),
    Symbol(&'src str),
    RightParenthesis,
    LeftParenthesis,
}

#[derive(Debug)]
pub struct Token<'src> {
    pub kind: TokenKind<'src>,
    pub span: SimpleSpan,
}

fn is_reserved_char(c: &char) -> bool {
    r#"(),;[]`{}_:"'"#.chars().any(|reserved| reserved == c.to_char())
}

#[must_use]
fn symbol<'a, I: StrInput<'a, C>, C: Char, E: ParserExtra<'a, I>>(
) -> impl Parser<'a, I, &'a C::Str, E> + Copy + Clone {
    let f = |c: &C| {
        let c = c.to_char();
        !c.is_whitespace() && !c.is_alphanumeric() && !is_reserved_char(&c)
    };

    any().filter(f).then(any().filter(f).repeated()).to_slice()
}

pub type LexerI<'a> = &'a str;
pub type LexerO<'a> = Vec<Token<'a>>;

#[must_use]
pub fn lexer<'s, E: ParserExtra<'s, LexerI<'s>>>() -> impl Parser<'s, LexerI<'s>, LexerO<'s>, E> {
    let number = text::int(10).map(|s: &str| TokenKind::Number(s.parse().unwrap()));

    let string = any()
        .filter(|c| *c != '"')
        .repeated()
        .to_slice()
        .map(TokenKind::String)
        .delimited_by(just('"'), just('"'));

    let right_parens = just(')').map(|_| TokenKind::RightParenthesis);
    let left_parens = just('(').map(|_| TokenKind::LeftParenthesis);

    let symbol = symbol().to_slice().map(TokenKind::Symbol);

    let ident = chumsky::text::unicode::ident().map(TokenKind::Ident);

    number
        .or(right_parens)
        .or(left_parens)
        .or(symbol)
        .or(string)
        .or(ident)
        .padded()
        .map_with(|t: TokenKind, e| Token {
            kind: t,
            span: e.span(),
        })
        .repeated()
        .collect()
}

#[cfg(test)]
mod tests {
    use super::*;
    use chumsky::error;
    use insta::assert_debug_snapshot;
    use rstest::rstest;
    use tracing_test::traced_test;

    type TestExtra = extra::Err<error::Cheap>;

    #[rstest]
    #[traced_test]
    fn test_symbol(
        #[values(
            //-
            " + ",
            " + ",
            " +2",
            " ++ ",
            " // ",
            " - - - -- - - ",
            " () , . ",
        )]
        input: &str,
    ) {
        let p = symbol::<_, _, TestExtra>()
            .padded()
            .repeated()
            .collect::<Vec<_>>();

        assert_debug_snapshot!(input, (input, p.parse(input)));
    }

    #[rstest]
    #[traced_test]
    fn test_lexer(
        #[values(
            ("empty", ""),
            ("int", "1 23 313 1"),
            ("symbols", "+ == != -"),
            ("symbols_split", r#"+3+1--3//3&(s++)"#),
            ("string", r#" "foo" "bar" "#),
            ("bad s", r#" "foo "#),
            ("parens", r#"(12 +23)()("foo")(1+1)"#),
            ("ident", "foo bar foo_bar foo-bar (foo+1)"),
        )]
        input: (&str, &str),
    ) {
        let p = lexer::<TestExtra>().padded();

        assert_debug_snapshot!(input.0, (input.0, input.1, p.parse(input.1)));
    }
}
