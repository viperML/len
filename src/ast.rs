use crate::lexer::TokenKind;
use crate::Int;
use chumsky::extra::ParserExtra;
use chumsky::pratt::{infix, left, prefix};
use chumsky::prelude::*;
use chumsky::Parser;
use std::ops::Not;

#[derive(Debug, Clone)]
pub enum Ast {
    Literal(Literal),
    FunctionCall(FunctionCall),
    Identifier(Identifier),
}

#[derive(Debug, Clone)]
pub enum Literal {
    Integer(Int),
    String(String),
    Boolean(bool),
}

#[derive(Debug, Clone)]
pub struct FunctionCall {
    pub(crate) function: Box<Ast>,
    pub(crate) argument: Box<Ast>,
}

#[derive(Debug, Clone)]
pub struct Identifier {
    pub(crate) name: String,
}

#[derive(Debug, PartialEq)]
pub struct Spanned2<T>(T, SimpleSpan<usize>);

fn is_infix(ident: &str) -> bool {
    ident
        .chars()
        .next()
        .unwrap_or_default()
        .is_alphabetic()
        .not()
}

pub fn expression_parser<'src, E: ParserExtra<'src, &'src [TokenKind<'src>]>>(
) -> impl Parser<'src, &'src [TokenKind<'src>], Ast, extra::Err<Rich<'src, TokenKind<'src>>>> {
    recursive(|expr| {
        let literal = select! {
            TokenKind::Ident("true") => Literal::Boolean(true),
            TokenKind::Ident("false") => Literal::Boolean(false),
            TokenKind::Number(x) => Literal::Integer(x),
            TokenKind::String(x) => Literal::String(x.to_string()),
        }
        .map(Ast::Literal);

        let ident = select! {
            TokenKind::Ident(x) => x,
        }
        .filter(|s| is_infix(s).not())
        .map(|s| {
            Ast::Identifier(Identifier {
                name: s.to_string(),
            })
        });

        let grouping = expr.clone().delimited_by(
            just(TokenKind::LeftParenthesis),
            just(TokenKind::RightParenthesis),
        );

        let atom = literal.or(ident).or(grouping).labelled("non-infix atom");

        let infix_ident = select! {
            TokenKind::Ident(s) => {
                if is_infix(s) {
                    Some( Ast::Identifier(Identifier {
                        name: s.to_string(),
                    }))
                } else {
                    None
                }
            },
        }
        .filter(|ast| ast.is_some())
        .map(|ast| ast.unwrap());

        atom.clone().pratt((
            prefix(3, atom, |op: Ast, o: Ast| {
                Ast::FunctionCall(FunctionCall {
                    function: Box::new(op),
                    argument: Box::new(o),
                })
            }),
            infix(left(2), infix_ident, |left: Ast, op: Ast, right: Ast| {
                let first_op = Ast::FunctionCall(FunctionCall {
                    function: Box::new(op),
                    argument: Box::new(left),
                });
                Ast::FunctionCall(FunctionCall {
                    function: Box::new(first_op),
                    argument: Box::new(right),
                })
            }),
        ))
    })
}

#[cfg(test)]
mod tests {
    use super::*;
    use chumsky::error;
    use insta::assert_debug_snapshot;
    use rstest::rstest;
    use tracing_test::traced_test;
    type TestExtra = extra::Err<error::Cheap>;

    #[rstest]
    #[traced_test]
    fn test_expression<'src>(
        #[values(
            ("empty", &[][..]),
            ("ident", &[TokenKind::Ident("foo")][..]),
            ("int", &[TokenKind::Number(1.into())][..]),
            ("app", &[
                TokenKind::Ident("foo"),
                TokenKind::Number(1.into())
            ][..]),
            ("app_multiple", &[
                TokenKind::Ident("foo"),
                TokenKind::Ident("foo"),
                TokenKind::Ident("foo"),
                TokenKind::Ident("foo"),
            ][..]),
            ("grouping", &[
                TokenKind::LeftParenthesis,
                TokenKind::Ident("foo"),
                TokenKind::RightParenthesis,
            ][..]),
            ("bad-grouping", &[
                TokenKind::LeftParenthesis,
                TokenKind::Ident("foo"),
            ][..]),
            ("grouping_app", &[
                TokenKind::Ident("foo"),
                TokenKind::LeftParenthesis,
                TokenKind::Ident("bar"),
                TokenKind::Ident("baz"),
                TokenKind::RightParenthesis,
            ][..]),
            ("grouping_app_alt", &[
                TokenKind::LeftParenthesis,
                TokenKind::Ident("foo"),
                TokenKind::Ident("bar"),
                TokenKind::RightParenthesis,
                TokenKind::Ident("baz"),
            ][..]),
            ("infix", &[
                TokenKind::Ident("foo"),
                TokenKind::Ident("+"),
                TokenKind::Ident("bar"),
            ][..]),
        )]
        input: (&str, &[TokenKind<'src>]),
    ) {
        let p = expression_parser::<TestExtra>();

        assert_debug_snapshot!(input.0, p.parse(input.1));
    }
}
