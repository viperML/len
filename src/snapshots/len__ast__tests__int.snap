---
source: src/ast.rs
expression: p.parse(input.1)
---
ParseResult {
    output: Some(
        Literal(
            Integer(
                1,
            ),
        ),
    ),
    errs: [],
}
